index.bs

```diff
@@ -6583,6 +6741,363 @@ This extension enables use of a user verification method.
+    dictionary AuthenticationExtensionsDevicePublicKeyInputs {
+        DOMString attestation = "none";
+        sequence<DOMString> attestationFormats = [];
+    };
+
+    partial dictionary AuthenticationExtensionsClientInputs {
+        AuthenticationExtensionsDevicePublicKeyInputs devicePubKey;
+    };
+    </xmp>
+    <div dfn-type="dict-member" dfn-for="AuthenticationExtensionsDevicePublicKeyInputs">
+        :   <dfn>attestation</dfn>
+        ::  The [=[RP]=] MAY use this OPTIONAL member to specify a preference regarding [=attestation conveyance=].
+            Its value SHOULD be a member of {{AttestationConveyancePreference}}.
+            [=Client platforms=] MUST ignore unknown values, treating an unknown value as if the [=map/exist|member does not exist=].
+
+            The default value is {{AttestationConveyancePreference/none}}.
+
+        :   <dfn>attestationFormats</dfn>
+        ::  The [=[RP]=] MAY use this OPTIONAL member to specify a preference regarding the [=attestation=] statement format used by the [=authenticator=].
+            Values SHOULD be taken from the IANA "WebAuthn Attestation Statement Format Identifiers" registry [[!IANA-WebAuthn-Registries]] established by [[!RFC8809]].
+            Values are ordered from most preferable to least preferable.
+            This parameter is advisory and the [=authenticator=] MAY use an attestation statement not enumerated in this parameter.
+
+            The default value is the empty list, which indicates no preference.
+    </div>
+
+    Note: To request the `devicePubKey` extension processing with default options, pass an empty dictionary as the input.
+
+: Client extension processing
+:: If {{AuthenticationExtensionsClientInputs/devicePubKey}} is present, the client creates the authenticator extension input from the client extension input.
+
+: Client extension output
+:: An ArrayBuffer containing the [=device public key attestation object=] that was returned in the authenticator extension output and the signature returned as the [=unsigned extension output=].
+    <xmp class="idl">
+    dictionary AuthenticationExtensionsDevicePublicKeyOutputs {
+        ArrayBuffer authenticatorOutput;
+        ArrayBuffer signature;
+    };
+
+    partial dictionary AuthenticationExtensionsClientOutputs {
+        AuthenticationExtensionsDevicePublicKeyOutputs devicePubKey;
+    };
+    </xmp>
+
+
+: Authenticator extension input
+:: A CBOR expression of the client extension input
+
+    ```
+    devicePublicKeyInputs = {
+      attestation: "none" / "indirect" / "direct" / "enterprise",
+      attestationFormats: [tstr],
+    }
+    $$extensionInput //= (
+      devicePubKey: devicePublicKeyInputs,
+    )
+    ```
+
+: Authenticator extension output
+:: A byte string containing the CBOR encoding of the <dfn>device public key attestation object</dfn>, defined by the `attObjForDevicePublicKey` type:
+
+    ```
+    $$extensionOutput //= (
+        devicePubKey: bstr,
```

Comment:
It looks to me like the `attObjForDevicePublicKey` shouldn't need to be conveyed as an opaque byte string, does it? We're not signing over `attObjForDevicePublicKey` in isolation, and when we do sign over it it's as part of the authenticator data, which is already conveyed as an opaque byte string, so we have a consistent encoding already. So there seems to be no reason we can't just declare the type as `devicePubKey: attObjForDevicePublicKey`, is there?

Reply:
Hm, I suppose that for the binary equality check on `get()`, you could try to match the entire `attObjForDevicePublicKey` byte string instead of each individual field. But right now the RP processing steps say to store and compare the fields individually.

Reply:
I switched this in https://github.com/w3c/webauthn/pull/1663/commits/ec03d4da8d88cd797a3c003c0212cc1beb438f4e . The reason is that the authenticator output is mirrored at the WebAuthn level, inline with other extensions and Mike Jones' general rule. But if the CBOR wasn't wrapped in a bytestring here, platforms would either have to re-encode it—which opens questions about whether it'll be bitwise equal—or else the WebAuthn level client extension would have to mirror the whole structure of the DPK, which seems like a lot of duplication.

Reply:
Ah, I see. I found our previous discussion here: https://github.com/w3c/webauthn/pull/1663#discussion_r930202660.

Though that does make me wonder: who might benefit from this duplication? This way the value is still a CBOR map, so unlike `getPublicKey()` you're not getting away from CBOR by reading it out of the client outputs instead of the authenticator outputs. That would be different if we say to transcode the whole data structure into JS objects, but I agree that seems like too much duplication. And in both cases you can't verify any signature over the client output without also parsing the authenticator data anyway.

I don't think it's a blanket rule that client outputs should try to mirror the authenticator output, since we do define that:

>Extensions [...] MUST define the process by which [...] the [CBOR](https://pr-preview.s3.amazonaws.com/w3c/webauthn/pull/1663.html#cbor) [authenticator extension output](https://pr-preview.s3.amazonaws.com/w3c/webauthn/pull/1663.html#authenticator-extension-output), and the [unsigned extension output](https://pr-preview.s3.amazonaws.com/w3c/webauthn/pull/1663.html#unsigned-extension-outputs) if used, can be used to determine the [client extension output](https://pr-preview.s3.amazonaws.com/w3c/webauthn/pull/1663.html#client-extension-output).

So I would prefer to just drop the `authenticatorOutput` property from the client outputs, since I don't see who would be using it anyway. Unless I'm missing something?

Anyway, I don't feel strongly about it, I won't hold up the PR over it if you prefer to keep it as is.

Reply:
Discussed on the 2022-10-05 call, let's keep as is for now.